decidim-core/lib/devise/models/decidim_newsletterable.rb

```diff
@@ -0,0 +1,59 @@
+# frozen_string_literal: true
+
+module Devise
+  module Models
+    # Newsletterable adds all needed validations to acomplish GDPR standards
+    # in newsletter notification mailing
+    #
+    # Validates that all users with notifications activated have Opt-in in
+    # some time
+    module DecidimNewsletterable
+      def newsletter_opt_in_notify
+        unless newsletter_opt_in_valid?
+          set_newsletter_token!
+          Decidim::NewslettersOptInJob.perform_later(self, newsletter_token)
+        end
+      end
+
+      def newsletter_opt_in_validate
+        set_newsletter_opt_in! if newsletter_opt_in_at.blank?
+      end
+
+      protected
+
+      # Checks if Opt-in is valid or not
+      # An user with notifications activated, must have an Opt-in
+      def newsletter_opt_in_valid?
+        !newsletter_notifications? || newsletter_opt_in_at.present?
+      end
+
+      # Fill token value with some random value
+      # Removes opt_int timestamp
+      # Deactivate newsletter notifications
+      def set_newsletter_token
+        self.newsletter_notifications = false
+        self.newsletter_token = SecureRandom.base58(24)
+        self.newsletter_opt_in_at = nil
+      end
+
+      # Sets token and Opt-in values & saves
+      def set_newsletter_token!
+        set_newsletter_token
+        save
+      end
+
+      # Fill Opt-in value with current time &
+      # removes any token involved
+      def set_newsletter_opt_in
+        self.newsletter_opt_in_at = DateTime.current
```

Comment:
What about new users? Shouldn't they have this set if they check the box at registration?

Reply:
Also, shouldn't we use this field to filter the user who receive emails? Right now it's doing nothing

Reply:
Keep in mind that this functionality will be only executed once, so new users will not have it set.
To add it, I recomend to implement a double-step confirmation to receive newsletters, in registration and in notifications settings modification.

Reply:
But if new users don't have it how is this flag going to be used? Shouldn't new users also give the consent?

Reply:
This functionality is not present in this task. It's only to confirm that pre-GDPR users are OK to receive newsletters.

Reply:
Even if we merge this, which I'm not really sure since it doesn't prevent from sending newsletters to users that don't want to, what about new users? They won't have this flag set even if when they sign up they choose to receive newsletters.

Reply:
I agree with @oriolgual on this - If we are to merge this, we should probably filter out emails to approve users who haven't approved.

Still, this looks much like a hack that shouldn't be in the core. GDPR is already in effect, so we can't send emails users who hadn't agreed with newsletters anyway.

The field introduced in this patch is quite redundant, as the currently `newsletter_notifications` should suffice.

Reply:
OK, I'm going to make a list:

* *We need to send a new email to all the users that have marked the newsletter checkbox*

By now is a one-time task execution, that tracks and sends an email to all current users that have `newsletter_notifications` check marked.

* *We should keep track which users have clicked on this button on the email*

This is the reason to add `newsletter_opt_in_at` field. Yes, as you say it's redundant, but is the only way to know which an when users have Opt-in.

* *we should probably filter out emails to approve users who haven't approved.*

I'm agree with that, but this means to enhance registration and notifications settings form, to add a double-step confirmation for newsletter (It has not been requested in this task, yet).

**So**, asking directly to @decidim/product, I still have time to program that double-step confirmation via mail for all users that want to receive newsletters. Current devs are not affected by this enhancement. Is this OK?

Reply:
Just to be clear, basically, what we need to solve it's that until now on [sign up form](http://staging.decidim.codegram.com/users/sign_up) it's not valid to send these emails on newsletters, as it's checked by default. 

So, we need to have an option to migrate these users so they opt-in themselves. 

> Still, this looks much like a hack that shouldn't be in the core. 

Maybe as it's a one use thing we could make an engine with this one? FYI, almost all installations already have this issue. 

> GDPR is already in effect, so we can't send emails users who hadn't agreed with newsletters anyway.

IANAL but I've talked with an Expert on Privacy and Data Protection Law, and she said that we could send these kind of mailing, but it should be clear on the releases notes that this decision should be made by a lawyer.

Reply:
@andreslucena the sign up form was already fixed on `master`, but maybe the staging app is not being updated. See #3488 and #3483.

Reply:
Maybe I got something wrong or I'm not explaining myself. I understand that this is a one-time task that has to be run for users that have `newsletter_notifications` set to true. Let me expose my issues:

* If we add this field but it's not used anywhere, it doesn't prevent admins from sending newsletters to users that haven't accepted it, so we aren't fixing any problem.

* If we deploy this, run the script, and wait somedays to add more features, what happens to the users that signed up during these days? Should we also send them a new email? (It doesn't make any sense)

* Why can't we update the registration process to just use the new field instead of the old one? I  don't see why we need a two step process. I understand that is enough if a users checks the box accepting to receive newsletters, there's no need to ask them again.

* Also, after talking with @mrcasals, we could drop the `newsletters_notifications` field at the database. Then, we can set the new field to a date prior to May 25th for the users that had `newsletters_notifications` to true and reset it after sending this one time mail.

Reply:
Keeping in mind that, as @andreslucena says, this is an optional feature, some questions:

* What about organizations that do not want to use this feature?
* It's necessary to know exactly when an user has Opted-in?

Reply:
> What about organizations that do not want to use this feature?
I was thinking on making it a rake task, so it isn't send by mistake. We should also ask for confirmation when we send it, for instance: 

```
$ rails decidim:gdpr:send_emails
This will send an email to all the XXX users that have marked the newsletter by default. This should only be run if you were using Decidim before v0.11
If you have any doubts regarding this feature, please check the releases notes for this version  https://github.com/decidim/decidim/releases/tag/v0.12
Are you sure you want to do that? [y/N]  
```

> It's necessary to know exactly when an user has Opted-in?
I don't understand the question, let's clarify this on private

Reply:
As @andreslucena says, it's usefull to keep the `newsletter_opt_in_at` timestamp field, to know when an user has click on e-mail button. And removing `newsletter_notifications` field will need more changes (models, rspec, migrations....)